#Quality Check Your Samples and Replicates

Once you've performed transcript quantification for each of your biological replicates, it's good to examine the data to ensure that your biological replicates are well correlated, and also to investigate relationships among your samples.  If there are any obvious discrepancies among your sample and replicate relationships such as due to accidental mis-labeling of sample replicates, or strong outliers or batch effects, you'll want to identify them before proceeding to subsequent data analyses (such as differential expression).  

The Trinity toolkit includes a script 'PtR' (pronounced Peter, and stands for Perl-to-R) that can be used to generate a variety of plots for exploring your matrix of expression data. We'll be using it in each of the steps below.  You'll be needing three files to proceed, two of which were [generated by the earlier abundance estimation routines](Trinity-Transcript-Quantification)

1. your TMM-normalized expression matrix, which we'll call 'TMM.EXPR.matrix'.

2. your fragment counts matrix, which we'll call 'counts.matrix'.

3. a 'samples.txt' file, with tab-delimited format:

      condition_A_name     replicate_A1_name
      condition_A_name     replicate_A2_name

      condition_B_name     replicate_B1_name
      condition_B_name     replicate_B2_name

and the replicate names should match identically to the column headers of your 'TMM.EXPR.matrix' and 'counts.matrix' files.


## Compare replicates for each of your samples

Run PtR like so to compare the biological replicates for each of your samples.

      % $TRINITY_HOME/Analysis/DifferentialExpression/PtR --matrix counts.matrix \
                      --samples samples.txt --CPM --log2 --compare_replicates

Here, PtR reads in the matrix of counts, performs a counts-per-million (CPM) data transformation followed by a log2 transform, and then generates a multi-page pdf file (named ${sample}.rep_compare.pdf) for each of your samples, including several useful plots:

A plot showing the sum of mapped fragments:

<img src="https://raw.githubusercontent.com/wiki/trinityrnaseq/trinityrnaseq/images/sum_of_mapped_fragments.png" width=400 />

Pairwise comparisons of replicate log(CPM) values. Data points more than 2-fold different are highlighted in red:

<img src="https://raw.githubusercontent.com/wiki/trinityrnaseq/trinityrnaseq/images/replicate_pair_scatter_plot.png" width=400 /> 

Pairwise MA plots (x-axis: mean log(CPM), y-axis log(fold_change)):

<img src="https://raw.githubusercontent.com/wiki/trinityrnaseq/trinityrnaseq/images/replicate_pair_MA_plot.png" width=400 /> 

and a replicate Pearson correlation heatmap:

<img src="https://raw.githubusercontent.com/wiki/trinityrnaseq/trinityrnaseq/images/replicate_pair_correlations.png" width=400 /> 
